url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/methodology/ParetoFrontier.java

```diff
@@ -0,0 +1,181 @@
+package methodology;
+
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.TreeMap;
+
+import protocols.specifications.DBProtocol;
+import utils.Progress;
+
+import attributes.Attribute;
+
+public class ParetoFrontier implements Serializable{
+	
+	private static final long serialVersionUID = -260614203005100896L;
+	
+	private DBProtocol[] protocols;
+	private DBProtocol[] frontier;
+	private TreeMap<Integer, List<Integer>> indexesToBeRemoved;
+	private Attribute[] attributes;
+
+	public Attribute[] getAttributes() {
+		return attributes;
+	}
+
+
+	public void setAttributes(Attribute[] attributes) {
+		this.attributes = attributes;
+	}
+
+
+	public ParetoFrontier(DBProtocol[] protocols, DBProtocol[] frontier, Attribute[] attributes, TreeMap<Integer, 
+			List<Integer>> indexesToBeRemoved){
+		this.protocols = protocols;
+		this.frontier = frontier;
+		this.indexesToBeRemoved = indexesToBeRemoved;
+		this.attributes = attributes;
+	}
+
+	
+	/*Trujillo- Apr 4, 2014
+	 * Compute the pareto frontier according to some attributes and order*/
+	public static ParetoFrontier computeParetoFrontier(DBProtocol[] protocols, Attribute[] attributes, InternalState state) throws FileNotFoundException, IOException{
+		TreeMap<Integer, List<Integer>> indexesToBeRemoved = state.indexesToBeRemoved;
+		Progress progress = state.progress;
+		for (int i = 0; i < protocols.length; i++) {
+			if (i <= state.i) continue;
+			for (int j = 0; j < protocols.length; j++) {
+				progress.addProgress();
+				if (!indexesToBeRemoved.containsKey(j)){ 
+					List<Integer> tmp = new LinkedList<>();
+					indexesToBeRemoved.put(j, tmp);
+				}
+				if (protocols[i].dominate(protocols[j], attributes)){
+					/*
+					System.out.println(protocols[i].getIdentifier()+" dominates "+protocols[j].getIdentifier());
+					History.printInfoOfDomination(protocols[i], protocols[j], attributes);
+					*/
+					indexesToBeRemoved.get(j).add(i);
+				}
+			}
+			progress.printProgress();
+			state.setCoordinates(i);
+			state.setProgress(progress);
+			state.setIndexesToBeRemoved(indexesToBeRemoved);
+			ObjectOutputStream out = new ObjectOutputStream(new FileOutputStream(state.identifier));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 70, 71. The resource is closed at line 72. There are other execution paths that do not contain closure statements, for example, when ObjectOutputStream.writeObject throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).